### PR TITLE
Clean up @field and @fieldparent usage

### PR DESCRIPTION
**Pull request by smowton**
* Centralise use of raw types and database predicates in FieldParent and FieldBase classes
* Deduplicate type predicates common to all fields
* Deduplicate predicates common to function parameters and results

### CODE REVIEW

File: ql/src/semmle/go/AST.qll [hunk 1]
```diff
@@ -140,6 +140,29 @@ class DeclParent extends @declparent, AstNode {
...
 }
 
+/**
+ * An AST node whose children include fields.
+ */
+class FieldParent extends @fieldparent, AstNode {
+  /**
+   * Gets the `i`th field of this node.
+   *
+   * Note that the precise indices of declarations are considered an implementation detail
```
Comment (max-schaefer) [inline ql/src/semmle/go/AST.qll:150]:
```suggestion
   * Note that the precise indices of fields are considered an implementation detail
```

File: ql/src/semmle/go/AST.qll [hunk 2]
```diff
@@ -140,6 +140,29 @@ class DeclParent extends @declparent, AstNode {
...
+   */
+  FieldBase getField(int i) { fields(result, this, i) }
+
+  /**
+   * Gets a child field of this node in the AST.
+   */
+  FieldBase getAField() { result = getField(_) }
+
+  /**
+   * Gets the number of child declarations of this node.
```
Comment (max-schaefer) [inline ql/src/semmle/go/AST.qll:161]:
```suggestion
   * Gets the number of child fields of this node.
```

File: ql/src/semmle/go/Decls.qll [hunk 3]
```diff
@@ -363,11 +363,10 @@ class TypeSpec extends @typespec, Spec {
 }
 
 /**
- * A field declaration in a struct type.
+ * A field declaration, of a struct, a function (in which case this is a parameter or result variable)
+ * or an interace (in which case this is a method or embedding spec)
```
Comment (max-schaefer) [inline ql/src/semmle/go/Decls.qll:367]:
```suggestion
 * or an interface (in which case this is a method or embedding spec)
```

**Comment by smowton** [comment]
@max-schaefer doh, fixed

File: ql/src/semmle/go/Decls.qll [hunk 4]
```diff
@@ -363,11 +363,10 @@ class TypeSpec extends @typespec, Spec {
 }
 
 /**
- * A field declaration in a struct type.
+ * A field declaration, of a struct, a function (in which case this is a parameter or result variable)
+ * or an interface (in which case this is a method or embedding spec)
```
Comment (max-schaefer) [inline ql/src/semmle/go/Decls.qll:367]:
```suggestion
 * or an interface (in which case this is a method or embedding spec).
```

File: ql/src/semmle/go/Decls.qll [hunk 5]
```diff
@@ -363,11 +363,10 @@ class TypeSpec extends @typespec, Spec {
 }
 
 /**
- * A field declaration in a struct type.
+ * A field declaration, of a struct, a function (in which case this is a parameter or result variable)
```
Comment (max-schaefer) [inline ql/src/semmle/go/Decls.qll:366]:
```suggestion
 * A field declaration, of a struct, a function (in which case this is a parameter or result variable),
```

File: ql/src/semmle/go/Decls.qll [hunk 6]
```diff
@@ -407,74 +415,66 @@ class EmbeddedFieldDecl extends FieldDecl {
...
-  /**
-   * Gets the function declaration to which this receiver belongs.
-   */
-  FuncDecl getFunction() { fields(this, result, _) }
+  ReceiverDecl() { fd.getField(-1) = this }
 
   /**
-   * Gets the expression representing the type of the receiver declared in this declaration.
-   */
-  Expr getTypeExpr() { result = getChildExpr(0) }
```
Comment (max-schaefer) [inline ql/src/semmle/go/Decls.qll:472]: Do we inherit this predicate from somewhere else now?

Comment (smowton) [reply ql/src/semmle/go/Decls.qll:472]: Inherited from FieldBase.

Comment (max-schaefer) [reply ql/src/semmle/go/Decls.qll:472]: :+1: Thanks for confirming.

**Pull request merged**